datastore/bolt/group.go

```diff
@@ -1,114 +1,134 @@
 package bolt
 
 import (
-	"bytes"
 	"context"
 	"encoding/json"
-	"fmt"
 
 	"github.com/frain-dev/convoy/datastore"
 	"go.etcd.io/bbolt"
 )
 
-const name string = "groups"
-
 type groupRepo struct {
-	db *bbolt.DB
+	db         *bbolt.DB
+	bucketName string
 }
 
 func NewGroupRepo(db *bbolt.DB) datastore.GroupRepository {
-	return &groupRepo{db: db}
-}
+	bucketName := "groups"
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte(bucketName))
+		return err
+	})
 
-func (g *groupRepo) FetchGroupsByIDs(ctx context.Context, ids []string) ([]datastore.Group, error) {
-	return nil, nil
+	if err != nil {
+		return nil
+	}
+
+	return &groupRepo{db: db, bucketName: bucketName}
 }
 
-func (g *groupRepo) LoadGroups(context.Context, *datastore.GroupFilter) ([]*datastore.Group, error) {
+func (g *groupRepo) LoadGroups(ctx context.Context, filter *datastore.GroupFilter) ([]*datastore.Group, error) {
 	var groups []*datastore.Group
 	err := g.db.View(func(tx *bbolt.Tx) error {
-		c := tx.Bucket([]byte(bucketName)).Cursor()
-
-		prefix := []byte(name)
-		var grpSlice [][]byte
-
-		for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
-			fmt.Printf("key=%s, value=%s\n", k, v)
-			grpSlice = append(grpSlice, v)
-		}
+		b := tx.Bucket([]byte(g.bucketName))
 
-		for i := 0; i < len(grpSlice); i++ {
+		return b.ForEach(func(k, v []byte) error {
 			var grp *datastore.Group
-			mErr := json.Unmarshal(grpSlice[i], &grp)
-			if mErr != nil {
-				return mErr
+			err := json.Unmarshal(v, &grp)
+			if err != nil {
+				return err
 			}
 
-			groups = append(groups, grp)
-		}
+			if len(filter.Names) > 0 {
+				for _, grpName := range filter.Names {
+					if grpName == grp.Name {
+						groups = append(groups, grp)
+					}
+				}
+			} else {
+				groups = append(groups, grp)
+			}
 
-		return nil
+			return nil
+		})
 	})
 
 	return groups, err
 }
 
 func (g *groupRepo) CreateGroup(ctx context.Context, group *datastore.Group) error {
-	return createUpdateGroup(g.db, group)
+	return g.createUpdateGroup(group)
 }
 
 func (g *groupRepo) UpdateGroup(_ context.Context, group *datastore.Group) error {
-	return createUpdateGroup(g.db, group)
+	return g.createUpdateGroup(group)
 }
 
 func (g *groupRepo) FetchGroupByID(ctx context.Context, gid string) (*datastore.Group, error) {
 	var group *datastore.Group
 	err := g.db.View(func(tx *bbolt.Tx) error {
-		id := name + ":" + gid
+		b := tx.Bucket([]byte(g.bucketName))
 
-		grp := tx.Bucket([]byte(bucketName)).Get([]byte(id))
+		grp := b.Get([]byte(gid))
 		if grp == nil {
-			return fmt.Errorf("group with id (%s) does not exist", gid)
+			return datastore.ErrGroupNotFound
 		}
```

Comment:
we can call this `buf` instead as that name denotes the byte slice property better